### ISSUE TITLE
Fixing MissingMethodException and related binding issues with FSharp.Core

### ISSUE THREAD

**Issue by abelbraaksma** [Microsoft/visualfsharp#789]
It has been reported on numerous places for different versions of F# (see for instance [this](http://stackoverflow.com/questions/23389752/missingmethodexception-when-running-a-unit-test-that-uses-fsharp-data), [this](https://github.com/fsprojects/FsUnit/issues/23) and [this](http://stackoverflow.com/questions/33105382/why-do-i-get-a-missing-method-exception-runtime-when-creating-a-sqlclient-type)).

The typical solution is to add an `app.config` file to the project with the proper `bindingRedirect` elements, but this only solves the most common cases.
- If you have a library assembly that is used by someone else, the `app.config` won't be used
- If you reference project files that each have an `app.config`, Visual Studio 2015 won't copy them to the output directory as `libraryname.dll.config`, leading to loading issues in your main application
- If your library depend on a third party library you can add a binding redirect yourself, but if your library is used by others, these will have to add the same binding redirects and so on.
- Single dll or single exe no-install shipment is not possible this way, not even if you create your own `AssemblyResolve` event, because the CLR won't allow you to redirect a strongly named assembly (returning FSharp.Core 4.4.0.0 when 4.3.1.0 was requested results in "wrong version error" in the Fusion Log Viewer)
- If your library is dynamically loaded, or shadow-copied-then-loaded, it may not find the `libraryname.dll.config` file, leading to very-hard-to-diagnose errors
- I see different behavior and different errors (and sometimes they disappear) depending on using RyuJIT or not
- Configuring automatic redirect binding fails in many practical cases.
- The result of all this is what I often see as "it works locally, but not remotely, so it's probably a bug in their system", leading to bugs like these, once locally resolved, rear their ugly head again when deployed, copied, moved to other systems etc. that may be configured slightly differently, leading to the error coming back.

This all feels very much like being back in "dll hell". I see a couple of possible improvements, but none solve this issue completely:
1. Let the FSharp installer create a GAC policy that does the redirect, this will then be system-wide, and this is the approach used by earlier editions of F# (at least that's how it looks in my registry). Though I may assume there is a reason this approach has been abandoned.
2. If the powers that build FSharp.Core have a way of bypassing the strong-name redirect problem, the `<Module>.cctor` of FSharp.Core could bind an event handler to `AppDomain.AssemblyResolve` which would then, upon finding any known version of F#, redirect it to self. It can in turn provide a callback or system function that allows programmers to override this behavior. It can also allow overrides in app.config. The business logic for this is relatively simple, the only technical drawback is: I doubt it is possible in the CLR ecosystem, esp. with respect to portability,.
3. If no full solution exists, let the F# compiler come up with a strong hint, or even an auto-generated `library.dll.config` for each file, based on the app.config, overriding the standard behavior of Visual Studio of not copying these files (this is the solution I am currently doing by hand, but see above for its major drawbacks).
4. Talk with the CLR team to fix the `AssemblyResolve` behavior so that we can create a solution ourselves. For instance, I currently have a Fody hack in place that binds to the `AssemblyLoad`  and `AssemblyResolve` events just to get the proper logging information for the (many!) times this goes wrong, but it is not sufficient to solve the issue completely.
5. Let the compiler report on assembly conflicts (would be nice if it could do a deep search, resolving each type) and suggest to resolve it by strongly-naming the and GAC-installing the assemblies. Not ideal and a lot of extra effort for each F# assembly, but at least a solution that works for all your _own_ assemblies.

I would much prefer at least some way where either the issue is auto-resolved every time everywhere (sounds hard), or a way that a programmer can concoct a solution _from code_ (as opposed to the much flawed `app.config` approach).

I realize that this is not an issue with F# alone, but so far I have the feeling it happens most with F#. in fact, for the several dozens or so projects I have seen since F# 4.4 came about and that I moved to .NET 4.5.2 or 4.6, only F# projects continue to have problems, some of them only [partially solvable](http://stackoverflow.com/questions/34222710/how-can-i-redirect-a-strongly-named-assembly-by-code-using-assemblyresolve-or-ot).

**Comment by KevinRansom** [comment]
As far as I know the current visual F# tooling does a pretty decent job of handling fsharp.core.versioning issues, unless there are bugs of course.

TypeProviders are the biggest versioning problem because we don't check the version of types that the TP hands back to us.  We need to make the compiler pay more attention to the types handed back, and validate them appropriately.

As for test harnesses built in F# running F# libraries built with a later version of F#, that is a tough scenario for the test harness authors to get right, because passing F# types back from the test to the harness usually requires unification.  Binding redirects are the only real solution to that.  However, it is possible to start an appdomain with a specific binding redirects, and so the test framework can run the tests in a new appdomain with the right redirects for each testing assembly.  It requires a lot more care but is probably the only workable solution.  I'm pretty sure that nunit does something similar.  For test cases built against earlier frameworks

The third issue is also a TP problem, caused by our laxness in validating returned types.

Getting Loader changes in the CLR into the CLR is probably going to be harder than anything you can imagine, for many reasons.  Over the last 15 years the CLR has learned that any change to the loader has a ton of unintended consequences,  most proposal will break many existing apps.  Other changes will make it harder to maintain compatibility going forward.

So let's list the scenarios that F# currently is broken for, and try to figure out what we need to do to solve each.

**Comment by abelbraaksma** [comment]
>  I'm pretty sure that nunit does something similar. For test cases built against earlier frameworks

Indeed it does (if you log the current app domain id from any test you get 2, not 0 in most cases), but the new app domain relies on the configuration file of the default app domain.

Also, the new app domain approach won't help users of your library, most programmers don't even know what an app domain is.

> So let's list the scenarios that F# currently is broken for, and try to figure out what we need to do to solve each.

 Consider:
* Create an F# library assembly (dll) for FSharp.Core 4.4, LibA.dll
* Reference FParsec.dll (build agains FSharp.Core 4.3)
* Create another library dll (C#, F#, doesn't matter) that references your F# assembly, LibB.dll
* If each has an `app.config`, the compiler will dump into the bin-dir `app.config` and `LibB.dll.config`
* Package or zip it and give it to some other user
* In his exe-project ProjX.exe he references LibA.dll
* He compiles and ships, the bind of ProjX.exe will have in the bin no `app.config`, or the `app.config` for the exe if he created one. Even if the auto-bindingredirect is on, it will _not_ create a redirect for `FParsec.dll` because it is not recognized as such. 

Result, unless the user explicitly checks the linking chain and finds out that FParsec.dll references an older version _and_ he finds out how to fix it (ProjX.exe.config), he _may_ get MissingMethodException (it won't always fire). He also may not, in which case he may find a hard-to-analyze issue after he ships.

This could be prevented if the compiler simply forbids having a (linked) reference to both version of FSharp.Core.

Today I've finally found a possible workaround that can be written in code (as opposed to having each assembly getting its own config file), but it is a little hackery, it relies on calling a private method of the app domain to force it to look again at the configuration file (which you then create "on the fly"). It is tricky to get right, because if the `<Module>.cctor` references _any_ method that is in another (System or otherwise) assembly, this will trigger the loader, leaving you with less options, because most handlers here behave like a singleton: once initialized there's no changing it.

The more idiomatic way would be to make `AssemblyResolve` work when I return the FSharp.Core 4.4 assembly when the CLR requests the FSharp.Core 4.3 assembly. I am still at a loss why this is declined, as the documentation I could find says the opposite. But changing this would mean to change the CLR and as you say, this is unlikely to happen.

Since I believe the `MissingMethodException` is cause by the following:
* Assembly indirectly links both 4.3 and 4.4 FSharp.Core
* A method called from FParsec assembly is compiled as a 4.4 method (not sure why)
* The JIT tries to find it in 4.3 assembly and fails

If that is true, another way to solve this is at the F# compiler level, to somehow make sure that if multiple side-by-side FSharp.Core assemblies are referenced, that a 4.4 method never binds incorrectly to a 4.3 method and v.v.

I will try to add a minimal example solution.

**Comment by dsyme** [comment]
> I will try to add a minimal example solution.

That would be useful. It sounds like some kind of bug in auto-binding-redirects (which I assume are being used)

**Comment by abelbraaksma** [comment]
> It sounds like some kind of bug in auto-binding-redirects (which I assume are being used)

Auto-binding redirects simply creates an `app.config` at compile time with the correct bindings. The problem explained here (sorry if I wasn't clear) is that that `app.config` solution does not work in a variety of cases leading to issues with FSharp.Core binding, which the CLR does not allow you to hand-bind (using `AssemblyResolve` event) to a different, higher version of itself if you attempt to resolve it from code.

Still working on a minimal reproducible example solution (not difficult or so, but time is consumed by generalizing my currently, surprisingly, working workaround).

**Comment by dsyme** [comment]
Ok, I'd like to see the repro. Also I'd like to understand if a dependency manager like Paket would have helped detect the inconsistency at an earlier stage (I've noticed few problems with DLL versions in projects using Paket)

**Comment by abelbraaksma** [comment]
[FSharpBug#789.zip](https://github.com/Microsoft/visualfsharp/files/70093/FSharpBug.789.zip)

Sorry that this took quite a while longer than I had hoped, as it required both a bit more research and resolving related issues in the projects where these issues manifested themselves to begin with.

The zip contains a solution file with all necessary projects (5 in total) to see the issue appearing. For best testing, it helps if you have FSharp.Core 4.3.0.0 in the GAC, and from tests *not* in the GAC. 

Instructions can be found in `SolutionInfo.txt`. The three NUnit test projects have each a `FusionLog` dir with the results of `fuslogvw`. On the top of each of these files is a repro-instruction (varying from "just run in NUnit GUI" to more elaborate tests).

One issue has been resolved: I stated above that it is not possible to use the `AssemblyResolve` to return a different version of a signed assembly. This is not true. If both versions are signed, you can return a different version (but even if your resolver is successful, it will be logged as failure in `fuslogvw`, hence the confusion). It is *not* possible to return a non-signed assembly when a signed assembly is requested, but that is not the issue here.

The tests show a one-level deep dependency. Matters become more complex and harder to resolve if you reference multiple assemblies that themselves reference older versions of FSharp.Core. If such an assembly is loaded prior to your `AssemblyResolve`, you are out of luck, as once a load fails, it will always fail (failure is cached by the runtime). Similar problems arise if you have 4.3.0.0 (FParsec, for instance) and 4.3.1.0 (FsUnit, for instance) indirectly referenced.

At the same time I have worked on a more solid solution. The `AppDomain.CurrentDomain.SetupInformation` returns an `AppDomainSetup` object, but this is a clone of the internal `_FusionStore` field, so any change has no effect. After a bit of digging, I found out that:
* You can set certain values using reflection on the private methods
* After you set these fields, you can call the internal `SetupFusionStore` method, which refreshes *only those parts that have been changed*, which comes in handy.

Essentially, the two methods that are required (not in this solution) to solve this by overriding the internal FusionStore with a new `ConfigurationFile` that was created on the fly are:

    type Try<'a> =
    | Success of 'a
    | Fail of Exception

    type private Try() =
        static member inline map f t = 
            match t with
            | Success s -> try Success (f s) with e -> Fail e
            | Fail e -> Try.Fail e

    let inline private safe f =
        try f() |> Try.Success
        with e -> e |> Try.Fail

    // call this to get the internal, non-cloned and updateable AppDomainSetup
    let inline private getFusionInternal() = 
        fun() -> appDomainType().GetProperty("FusionStore", BindingFlags.GetProperty ||| BindingFlags.NonPublic ||| BindingFlags.Instance)
        |> safe
        |> Try.map (fun t -> t.GetValue(AppDomain.CurrentDomain) :?> AppDomainSetup)

    // call this after you updated the AppDomainSetup, with a clone of the original as first arg
    // it will only update changed settings (i.e., ConfigurationFile)
    let inline private updateFusionInternal oldFusion newFusion =
        fun() -> appDomainType().GetMethod("SetupFusionStore", BindingFlags.InvokeMethod ||| BindingFlags.NonPublic ||| BindingFlags.Instance)
        |> safe
        |> Try.map (fun t -> newFusion |> Option.map (fun newf ->  t.Invoke(AppDomain.CurrentDomain, [|newf;oldFusion|]) |> ignore))

I doubt very much that this is of use here (if so it should be configurable and preferably be part of `FSharp.Core.<Module>.cctor`, or any of the startup methods that F# injects), but if this can be used, I am willing to come with a more elaborate solution.

I currently can only use this with my own libraries, but it still won't work 100% (in particular, when my own libraries are not the first to be loaded). By adding it to FSharp.Core 4.4.0.0, which *will* be loaded prior to any other F# library version (if your main project is build for that version), it may fix the majority of issues, but I would expected some frowning upon hacky nature of this fix.

**Comment by abelbraaksma** [comment]
@KevinRansom and @dsyme, have you been able / had the time to check the repro?

**Comment by dsyme** [comment]
@abelbraaksma I looked at Test.Parse and repro'd the ``Could not load file or assembly 'FSharp.Core`` error.  However it seems like you're simply missing an App.config for that project.  Adding an App.config in Test.Parse with an appropriate binding redirect (see below) made the tests pass ok.  As such I don't yet see a bug here.  I didn't look at the rest of the repro since I suspect the missing App.config for the tests is the source of the problems - if not please link to an updated repro.

FWIW the binding redirect didn't get auto-generated for the Test.Parse project even if there is a near-empty App.config - I had to put the binding redirect in explicitly.  I'm not sure if that binding redirect is meant to be autogenerated or not for a test DLL.  But regardless it can be specified manually.

```
<?xml version="1.0" encoding="utf-8" ?>
<configuration>
    <startup> 
        <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.5.2" />
    </startup>
	<runtime>
		<assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
			<dependentAssembly>
				<assemblyIdentity name="FSharp.Core" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
				<bindingRedirect oldVersion="0.0.0.0-4.4.0.0" newVersion="4.4.0.0" />
			</dependentAssembly>
		</assemblyBinding>
	</runtime>
</configuration>
```

**Comment by abelbraaksma** [comment]
> However it seems like you're simply missing an App.config for that project

That's the whole essence of the issue. When you create an executable, VS will copy the app.config to executablename.exe.config and all is fine. If you create a library project, it will do the same *for that library only*. When you add that library project to another project, that other project *will not have these binding redirects*.

That is the point I was trying to make with the repro. So, whenever a programmer creates a library project (a dll) that internally uses another library that was compiled against a previous version of `Fsharp.Core`, it will not be possible to have it "run out of the box".

This can be fixes by FSharp.Core itself, possibly configurable by an assembly attribute on the assembly that is dependent on a higher version. FSharp.Core can then prevent that older versions of itself to be loaded within the same appdomain.

In short: if `A.dll` references `Fsharp.Core` 4.4, and references `B.dll` which itself references `FSharp.Core` 4.3, then even if `A.dll` has a binding redirect, this will *not* become part of any new project that references `A.dll`. 

Since I have only seen hard-to-debug and hard-to-understand errors when one appdomain loads two versions of `Fsharp.Core` I think it is safe to default to loading *one and only one version of `FSharp.Core`* within any appdomain. There are several ways this can be achieved.

**Comment by abelbraaksma** [comment]
Note that some other system libraries behave similarly. For instance, if you reference two versions of `mscorlib` (and I think also `System.Core`, but didn't check), only the high version will get loaded. So internally, at least for some assemblies, Microsoft has done something similar.

**Comment by abelbraaksma** [comment]
Also: there are cases where a simple bindingredirect will not work: if you load an assembly dynamically. In these cases, having an `AssemblyResolve` event will not help, because Fusion finds the GAC entry of the older strongly named assembly first. Only binding redirects can override that (which, unless you use the fix proposed, cannot be changed at runtime), so that has to be arranged on a different level, hence my suggestion.

**Comment by dsyme** [comment]
@abelbraaksma Your Test library which was missing the App.config is a set of tests which get executed and is never referenced by other applications.   As discussed in http://fsharp.github.io/2015/04/18/fsharp-core-notes.html, dlls like this containing tests (and also dlls which are addins) are effectively executable applications and need an App.config to work correctly.

In the .NET design, libraries do not get their own "composable" binding redirect config of the kind you suggest - only applications get binding redirect config.   That's not an F# decision. 

I don't think it's feasible for F# to take a different decision than .NET here so this is probably best discussed in .NET forums - unless there's an actionable bug per CONTRIBUTING.md. That said, I know there's lots of pain around binding redirects with .NET (and with FSharp.Core specifically since it is out-of-framework and referenced by many components), it's one reason I wrote the guide above.  

BTW if you see ways to improve that guide please do submit a PR to the guide, it would be most welcome.

**Comment by abelbraaksma** [comment]
@dsyme, I'm aware of how binding redirects are supposed to work. For a single application, VS will create at most one `app.config` and copies it to the output dir, renaming it to `destinationname.[exe|dll].config`. But it will not do this for projects that reference another project that itself has a binding redirect.

I agree that this is not only an F# issue, esp. in the sense that this behavior is "by design" per how .NET works here. But I think it becomes an F# issue because the core library does not entirely behave as a "normal" library and so many people are having issues with this and it so easily goes wrong.

*If* we can / want to do something about this, I think we do have a few options without violating existing behavior:

0. Let the F# compiler warn about this when building a project, or better yet, raise an error. I prefer an error, because this issue leads to unrecoverable runtime issues. Preferably other .NET compilers should do the same, but I doubt we can influence that.
0. Implement my suggestion, which means simply to write a small function that runs on `Fsharp.Core`'s `<Module>.cctor` which fixes this whole issue at root level. This solution should only be activated if no app.config is present (easily checked through code).
0. Make either or both optional as an argument for `fsc.exe` and/or an assembly-level attribute in the compiled project
0. Remove older versions from the GAC (only leave highest in) which allows programmers use an `AssemblyResolve` event to solve it themselves (but this may invalidate older versions)
0. Add a policy to existing GAC-registered versions, which is perhaps the "friendlier" way to resolve this. The policy will work the same way a binding redirect.

My fear is that if we do nothing, the "dll hell" -like issues will only grow with future versions. It also limits the marketability of F#-build libraries, because each and every potential user will have to be given instructions on how to use your product. It is not just "plug and play" as with "normal" (i.e., C# or VB.NET) lbiraries.

I think the only sure-fire way is to solve this in `FSharp.Core` itself, and we can if we want to. I'm willing to invest in writing a stable redirect/fusion fixer for this, but it is little use if I know beforehand it'll be denied.

**Comment by smoothdeveloper** [comment]
I think @abelbraaksma is making a strong case regarding vending F# libraries.

Since I've introduced F# in some of my projects at work, I've been hitting the issue a lot, while I never had serious trouble before (despite good amount of dependencies), it reminded me of issues I had in past upgrading Castle dependencies in some projects because Castle.Core was used in many third part dependencies beside my own code.

Even after I've learnt about the reasons it still feels to me that I'm going to hit those issues frequently and if FSharp.Core and / or F# compiler can have some mechanisms to reduce the headache that would be a great improvement.

**Comment by dsyme** [comment]
@abelbraaksma @smoothdeveloper FSharp.Core should be seen as a normal library.

My understanding is that the Visual F# Tools team want this to be more and more the case going forward, e.g. to deliver FSharp.Core as a nuget package like many .NET components.  They have decided to aim to reduce the reliance on machine-wide installation of FSharp.Core in the GAC, and have long recommended requiring copy-local of FSharp.Core like every other normal library. @KevinRansom and @otawfik-ms  can give more details.    I don't see an appetite in the Visual F# Tools team for F#-specific solutions in this area, and I fully understand why.

From what I've seen, problems for users are caused by:

- mistakenly relying on having FSharp.Core in the GAC (or equally, people not shipping FSharp.Core as part of their application)

- mistakenly not using binding redirects at all

- incorrect templates (not including an app.config, or AutoGenerateBindingRedirects or not setting CopyLocal=true)

Adding the FSharp.Core nuget package to your project (or as a dependency of your nuget package) is also a great way to make things work well.  Initially we were unsure of recommending that, but it seems to be working very well for people. 

Best
Don

**Comment by abelbraaksma** [comment]
@dsyme, I understand if this is a "no-go" area in the sense of fixing things. I think it is mainly broken because earlier versions (up until and including FSharp.Core 4.4) have been registered in the GAC. 

None of the above fixes the issues I have raised here, which cannot be fixed with an `app.config`, a `AutoGenerateBindingRedirects` or setting `CopyLocal=true`. The guidance page suggests not bundling `FSharp.Core` with your library, as the host should choose. That is true, but if your library is dependent on another library, which itself is compiled for `FSharp.Core vOld`, it won't matter what the user of your library does, unless he is passes this instruction down the chain (that is: any application, even when they dynamically load your lib, _must_ have an app.config with a binding redirect).

My whole point here is: this is not the case in the real world. And since we are not raising exceptions when conflicting libraries are loaded (I think that is a bad thing), we suffer the effect of violating the **[Throw Early, Catch Late](http://www.wikijava.org/wiki/10_best_practices_with_Exceptions#Throw_early_catch_late)** principle of exception handling: end users are given a mysterious `MissingMethodException` which is virtually impossible in a type-safe world.

Even if we are not allowed to fix the binding (i.e., by preventing multiple versions of `FSharp.Core` in the same domain), we should try our best to make `FSharp.Core` behave like a good citizen.

**Comment by dsyme** [comment]
> ... any application, even when they dynamically load your lib, must have an app.config with a binding redirect ...

Yes this is the biggest pain point.   It seems the .NET philosophy is that applications should be loading extensions via mechanisms that support configuration for those extensions, e.g. loading the extensions into a new appdomain with config.  Then your extension as an ext.dll.config and everything works OK.  Or else using some even more complex system like MEF.

However that approach is a big problem when the  interface to the extension is a complex object model as it's not particularly easy to interact across the appdomain boundary, and can be expensive to do so.  Also  CoreCLR doesn't support appdomains....

**Comment by dsyme** [comment]
Closing this old discussion
